Credential Plugins

jakemcdermott opened the pull request:
- [x] plugin system
- [x] tooling
- [x] model validation
- [x] serializer, views, urls
- [x] functional tests
- [x] access.py
- [x] docs/credential_plugins.md
- integrations:
  - [x] azure kms
  - [x] cyberark aim
  - [x] cyberark conjur
  - [x] hashicorp signed ssh
  - [x] hashicorp kv lookup (v1 & v2)
- [x] ui 
- [x] fix and update e2e tests

##### SUMMARY
Other pull requests for this feature (already merged):
- https://github.com/ansible/awx/pull/3021
- https://github.com/ansible/awx/pull/3052
- https://github.com/ansible/awx/pull/3257

related issues:
- https://github.com/ansible/awx/issues/2238
- https://github.com/ansible/awx/issues/1654
- https://github.com/ansible/awx/issues/2834


##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API
 - UI

##### ADDITIONAL INFORMATION
![cred_plugins](https://user-images.githubusercontent.com/9753817/54286783-deee5a00-457a-11e9-85bb-831d03196cd0.gif)

jakemcdermott committed:
add credential plugin system and minimal working hashivault

ryanpetrello commented on awx/main/models/credential/__init__.py:
```diff
@@ -481,6 +498,7 @@ class Meta:
         ('scm', _('Source Control')),
         ('cloud', _('Cloud')),
         ('insights', _('Insights')),
+        ('credential_management_service', _('Credential Management Service')),
```
This key is a _liiiiiiittle_ wordy.  What about something like `external`?

jakemcdermott replied:
works for me - done.

ryanpetrello commented on requirements/requirements.txt:
```diff
@@ -56,44 +57,45 @@ jinja2==2.10
 requests-futures==0.9.7
 requests-oauthlib==0.8.0  # via social-auth-core
-requests[security]==2.15.1
+requests==2.15.1
```
What's this change about?

jakemcdermott replied:
Unsure - this is the file that was generated after following the steps in the readme

ryanpetrello replied:
Let's change this back then.

jakemcdermott replied:
alright - I'll change this back manually. I'll do the same for `docutils` - the auto-generator took that out.

ryanpetrello replied:
Hrrrrm cc @chrismeyersfsu was it you who added the `docutils`?

chrismeyersfsu replied:
If boto doesn't require it you can remove docutils.

My runner integration branch will install docutils via a Makefile directive _before_ the requirements are installed in order to satisfy `python-daemon`

jakemcdermott replied:
that makes sense - I'll go ahead and remove it, then.

jakemcdermott replied:
After digging into this a bit, I don't think the changes we are discussing here have much to do with the library I'm adding for the credential plugin backend so I've moved all of this to it's own PR: https://github.com/ansible/awx/pull/3111.

ryanpetrello commented on tools/docker-hashivault-override.yml:
```diff
@@ -0,0 +1,15 @@
+    cap_add:
+      - IPC_LOCK
+    environment:
+      VAULT_DEV_ROOT_TOKEN_ID: 'cupcakes'
```
![giphy](https://user-images.githubusercontent.com/214912/51921894-2eefd380-23b6-11e9-8a32-a48d0b0295e7.gif)

ryanpetrello commented on awx/api/serializers.py:
```diff
@@ -2768,6 +2768,20 @@ def create(self, validated_data):
                 if value:
                     validated_data['inputs'][attr] = value
+
+        # XXX - Temporary hack for dev and demonstration purposes
```
Before we merge, let's definitely replace this with new API views+serializers that let you manage `CredentialInputSource` objects.

jakemcdermott replied:
sounds good - will update with new views and serializers

jakemcdermott commented on awx/main/models/credential/__init__.py:
```diff
@@ -1227,3 +1256,45 @@ def tower(cls):
+        default=False,
+        editable=False,
+    )
+
```
I'm thinking we probably want some additional pre-save validation here: something that prevents a save and throws an exception if the target credential is an `external` credential or if the source credential anything but an `external` credential.

ryanpetrello replied:
Yep, that's true.

ryanpetrello commented:
@jakemcdermott another thing I'd like w/ this PR is some tests like these:

https://github.com/ansible/awx/blob/devel/awx/main/tests/functional/api/test_credential.py

...that use the new API view+serializer to associate/disassociate/confirm validation works properly and prevents things like 
https://github.com/ansible/awx/pull/3098#discussion_r251910300

ryanpetrello commented on requirements/requirements.in:
```diff
@@ -22,7 +22,7 @@ django-taggit==0.22.2
 djangorestframework-yaml==1.0.3
 irc==16.2
-jinja2=2.10
+jinja2==2.10
```
d'oh

jlmitch5 reviewed (dismissed):
On the UI side there doesn't seem to be any code changed

jakemcdermott commented:
@jlmitch5 good catch, thank you. - if UI changes end up in this PR I will re-tag it.

jakemcdermott committed:
add api for managing credential input sources

ryanpetrello commented on awx/api/urls/urls.py:
```diff
@@ -119,6 +120,7 @@
 v2_urls = [
     url(r'^$', ApiV2RootView.as_view(), name='api_v2_root_view'),
     url(r'^credential_types/', include(credential_type_urls)),
+    url(r'^credential_input_sources/', include(credential_input_source_urls)),
```
If there a reason to have these at a global level, or is `/api/v2/credentials/N/input_sources/` good enough for our needs?  When would you create/manage one of these _outside_ of the context of managing a credential?

ryanpetrello commented on awx/api/views/root.py:
```diff
@@ -101,6 +101,7 @@ def get(self, request, format=None):
         data['credentials'] = reverse('api:credential_list', request=request)
         if get_request_version(request) > 1:
             data['credential_types'] = reverse('api:credential_type_list', request=request)
+            data['credential_input_sources'] = reverse('api:credential_input_source_list', request=request)
```
Same question as above.

ryanpetrello replied:
Ignore this, I've pushed a commit to address it.

ryanpetrello commented on awx/main/access.py:
```diff
@@ -1111,6 +1111,19 @@ def can_delete(self, obj):
+     - I'm a superuser (TODO: Update)
+    '''
+
+    model = CredentialInputSource
```
There may not be much work here; most of the `BaseAccess` implementation actually _defaults_ to superuser-only.

ryanpetrello replied:
@AlanCoding do we actually need to _do_ anything here?  Don't these all just default to `superuser` only?

ryanpetrello committed:
switch CredentialInput creation to use the associate/disassociate view

ryanpetrello reviewed (commented):
This seems to work pretty well for me in my testing.  I tested this by spinning up the `docker-compose-hashivault` target, making a v1 kv engine and adding a username + password:

<img width="641" alt="image" src="https://user-images.githubusercontent.com/214912/52978361-f5036300-339e-11e9-994a-615306dd7d11.png">

Next, I added the credential in awx:

<img width="1193" alt="image" src="https://user-images.githubusercontent.com/214912/52978377-08163300-339f-11e9-8a25-69ce20d51a1c.png">

...and associated the input source to my Machine credential (no UI for this yet):

<img width="522" alt="image" src="https://user-images.githubusercontent.com/214912/52978391-1a906c80-339f-11e9-8551-0d3fae86ad0f.png">

```
In [39]: print(' '.join(json.loads(Job.objects.get(pk=10).job_args)))
/usr/bin/ansible-playbook -i /tmp/awx_10_ps5j93uk/tmpuw86_odl -u this-came-from-hashicorp-vault -e @/tmp/awx_10_ps5j93uk/tmp5110mgkd hello_world.yml
```

AlanCoding commented on awx/main/models/credential/__init__.py:
```diff
@@ -35,9 +40,10 @@
+__all__ = ['Credential', 'CredentialType', 'CredentialInputSource', 'V1Credential', 'build_safe_env']
 
 logger = logging.getLogger('awx.main.models.credential')
+credential_plugins = [ep.load() for ep in iter_entry_points('awx.credential_plugins')]
```
I should ask the obvious question of why `awx.credential_plugins`, as opposed to `awx.plugins.credentials`, although this might be more like a "credential input" plugin, because talking about credential plugins makes me think of credential types.

The answer is probably that `awx.plugins` is not an actual python module.

That gets to a larger issue about plugins, which I really do want to talk about. I know how notification template plugins are implemented, and it's not quite the direction that I would prefer to go in. Instead of looking at those as a model to start with building a new type of plugin system, I would look to:

https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/loader.py

In particular, see `imp.load_source`. I think that the Ansible core team has a ton of things they can teach us about doing this right. They are really the experts, and have spent a huge amount of resources building out a plugin system that is flexible and works for everybody.

I think what their implementation shows us is that plugins really shouldn't be loaded by default as python modules. I know notifications use some metadata from the inherited class to make some decisions, and it might be harder to make those decisions on-demand.

jakemcdermott replied:
> I should ask the obvious question of why `awx.credential_plugins`, as opposed to `awx.plugins.credentials`

No particular reason - naming stuff is hard. `plugins.credentials` works for me.

ryanpetrello replied:
`setuptools` entry points are built for _exactly_ this plugin model.  This approach is something that @matburt and I recommended to @jakemcdermott for this PR as lots of popular Python libraries use this approach as a mechanism for supporting and discovering third-party plugins.

Ansible core's plugin model is _fine_, but it's something _they_ invented.  I'd prefer to use something that's commonplace in the Python community and has a low barrier to entry, especially since many potential integrators will be third party software vendors that want an easy path to providing plugin integration.

matburt replied:
I think core's plugin model is grossly complex for our use case. The basic setuptools plugin entrypoints are the way to go. Notification Templates should have probably been implemented this way (and it's my fault they aren't)

jakemcdermott replied:
> The answer is probably that `awx.plugins` is not an actual python module.

Nope. You're right. It is not a module. It is a shared namespace for plugins.

ryanpetrello replied:
I _do_ like `awx.plugins.credentials`.  Might as well stick the landing on the namespace while we're here.

AlanCoding replied:
gotcha, I see the reasoning for using it in the entry_points.

AlanCoding commented on awx/main/models/credential/__init__.py:
```diff
@@ -275,6 +280,10 @@ class Meta:
 
+    def __init__(self, *args, **kwargs):
+        super(Credential, self).__init__(*args, **kwargs)
+        self.dynamic_input_fields = self._get_dynamic_input_field_names()
```
This is going to be a real major performance issue, because any time a list of credentials is loaded, which would have otherwise been `O(1)` queries, this will incur `O(N)` queries.

jakemcdermott replied:
Yeah - Instead of doing this, I was thinking about making the query the first time `get_input` is called and caching the field names for subsequent calls.

AlanCoding commented on awx/main/models/credential/__init__.py:
```diff
@@ -461,8 +472,18 @@ def get_input(self, field_name, **kwargs):
         return field_name in self.inputs and self.inputs[field_name] not in ('', None)
 
+    def _get_dynamic_input_field_names(self):
+        input_sources = CredentialInputSource.objects.filter(target_credential=self)
```
I think it's better to use the related name when possible, in this case, `input_source` I believe (it might be good to make this plural). That will use the prefetched objects if someone does `Credential.objects.prefetch_related('input_source')`. It might still use it in this case, but I think related name is better. That said, it still shouldn't be called from a model's `__init__`.

jakemcdermott replied:
Agreed on using the related names when possible. I _think_ the reason I bypassed them here was to get around some credential `__getattr__` behavior but I'll double-check. If the related names work we can definitely switch to using those.

AlanCoding replied:
It's possible that the entirety of this method (and the other one) could be reduced to

```python
    @cached_property
    def dynamic_input_fields(self):
        return [obj.input_field_name for obj in self.input_sources.all()]
```

ryanpetrello committed:
Improve the HashiCorp Vault KV name and field labels/help_text

ryanpetrello committed:
support HashiCorp Vault versioned secrets (API v2)

ryanpetrello committed:
give credential plugins an explicit namespace

ryanpetrello committed:
add credential plugin support for Azure Key Vault

ryanpetrello commented on requirements/requirements.in:
```diff
@@ -38,7 +39,7 @@ python-radius==1.0
 python3-saml==1.4.0
 social-auth-core==3.0.0
 social-auth-app-django==2.1.0
-requests<2.16  # Older versions rely on certify
```
@AlanCoding do you know why we pinned this?

https://github.com/ansible/awx/commit/f4a241aba238b2f64e5734f297325098d8e9eda8
https://github.com/ansible/awx/pull/1415

AlanCoding replied:
https://github.com/ansible/awx/pull/1198#issuecomment-365140621

Is pretty much all there is on background on that.

At the time, this was still an upgrade, because we had been at a lower version at like 2.15. So when I did it, I upgraded it just shy of 2.16. We certainly don't want to stay there. Obviously certify is added here. I would hope that any problems that may or may not have existed then won't still exist.

ryanpetrello replied:
Looks like `azure-keyvault` wants a newer version, so we'll have to see what falls out of this.

wenottingham replied:
probably some nonsense with certify and ceritifcate verification for LDAP

ryanpetrello committed:
add support for HashiCorp signed SSH certificates

jakemcdermott committed:
add api for testing credential plugins

ryanpetrello committed:
move path parameterization to the CredentialInputSource model

ryanpetrello committed:
don't add hvac as a dependency for hashicorp vault integration

hvac is just based on requests anyways, and it doesn't support half of
what we need (like the SSH secrets engine API)

ryanpetrello committed:
mark cred plugin strings for translation

ryanpetrello commented on awx/main/migrations/0062_v350_credential_plugins.py:
```diff
@@ -0,0 +1,53 @@
+class Migration(migrations.Migration):
+
+    dependencies = [
+        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
```
I want to revisit these before we merge and figure out why Django is auto-adding them.

ryanpetrello replied:
I tried recreating these by hand, and Django insists on adding them when we run `makemigrations`.  @matburt do you have enough Django-fu to guess why 🤷‍♂️ ?

jakemcdermott replied:
![](https://user-images.githubusercontent.com/9753817/53578660-0e966e80-3b46-11e9-9c49-e7a70d4380c4.gif)

ryanpetrello committed:
add RBAC definitions for CredentialInputSource

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -1418,6 +1418,80 @@ class CredentialCopy(CopyAPIView):
+    serializer_class = serializers.CredentialInputSourceSerializer
+
+
+class CredentialInputSourceSubList(SubListCreateAttachDetachAPIView):
```
It is easier for API clients to interact with global lists like `/api/v2/input_sources/` than with sublists like `/api/v2/credentials/N/input_sources/`. Creating one via the global list only means they have to provide both the source and target credential ids. Using the sub-list means they have to know which one is the "parent" resource, and this isn't always obvious to the person writing integrations, it's a distinction they would prefer to not have to think about. For that reason, I advocate introducing a global list as well. If there's any endpoint we don't need, I think it's the sublist.

ryanpetrello replied:
_Maybe_?  We sort of made the decision (for the sake of simplicity) to pick _only_ one of these to support, and the associate/disassociate model made sense to us here because the action you're taking is almost exclusively _linking_ and _unlinking_ two things (the global list view make users _care_ that an entity called an "input source" exists under the hood - I'd argue that they shouldn't have to know that detail).  I also don't love adding _another_ top level resource at `/api/v2/`, especially considering this one is *so* specific to the notion of "get Credential `A`'s field value `Y` from Credential `B`".

AlanCoding replied:
> I also don't love adding _another_ top level resource at `/api/v2/`, especially considering this one is _so_ specific to the notion of "get Credential `A`'s field value `Y` from Credential `B`".

Whenever @matburt talked about this, the principle was always that we make our API "as flat as possible". That's not an actual quote, but I hope I capture the spirit.

So I would respond saying almost the exact rhetorical opposite - `/api/v2/` is tremendously valuable as a comprehensive list of the resources in the system. If a model is not represented there it creates an unwelcome surprise. When the 1 other model I know of which wasn't listed in the top-level endpoint was written into tower-cli, we did very ugly hacks to work it in.

> the associate/disassociate model made sense to us here because the action you're taking is almost exclusively _linking_ and _unlinking_ two things

I agree completely with you for all of the "through" models. For instance, I do like that the group-host and group-children models are not exposed as top-level items. An even better example would be `RoleAncestorEntry`. We define this model explicitly and it has extra fields, but these are never exposed to API clients.

But this `CredentialInputSource` model has user-editable fields. For that matter, job templates are just expressing a relationship between projects and inventories (with a few extra user-editable fields).

ryanpetrello replied:
Should we support **creation** via _both_ mechanisms, then?  I think having a global _list_ is fine, but I ended up doing it this way because you're almost never going to _create_ one of these entities _outside_ of the context of attaching it to an existing credential. 

What if `/api/v2/credential_input_sources/` was just a read-only list?

AlanCoding replied:
> What if `/api/v2/credential_input_sources/` was just a read-only list?

I think that's the worst option of all. I would rather it not be added than to do that. The only thing I was trying to comment on is consistency. For the user, this means consistency of all resource types across the entire app. If this is merged with me having the minority opinion on this, then that's still fine.

Let me return to some other things here. The view class you're using is `SubListCreateAttachDetachAPIView`. I don't think that's entirely right.

How would you attach or detach here? Does that make any sense? When I read your RBAC logic, it implied that `target_credential` could never be null. If it can't be null, then how can you detach? Perhaps you want a sublist-create-API-view? There are 2 categories of sublists - many-to-many relationships and reverse foreignkeys, and they should generally inherit from different base classes. You can still do the creation for many-to-many views, but attach-detach doesn't make sense with reverse foreignkeys. If you were to detach or replace (replace, since this is non-nullable) a foreignkey relationship, that's a PATCH, not a POST {"id": 42} type pattern.

AlanCoding replied:
Anyway, it looks like the global list is implemented as read-only with `CredentialInputSourceList`

ryanpetrello replied:
@AlanCoding 

The disassociate in this test seems to work fine for me:

https://github.com/ansible/awx/pull/3098/files#diff-731de063081413220909cdc9f7bc3447R38

Am I missing something?

AlanCoding replied:
I think that test articulates exactly why you shouldn't be able to detach here.

The "disassociate" request is made.

```python
response = post(sublist_url, params, admin)
```

After that request completes, no credential input sources exist

```python
assert CredentialInputSource.objects.count() == 0
```

The act of disassociating the credential input source deleted the credential input source.

I assume this is due to the cascade delete behavior, which is what I would have expected. However, no sane user would ever expect this. If you want to delete it, you should use a DELETE request.

ryanpetrello replied:
Okay, I've added a commit which should address this feedback.

AlanCoding commented on awx/main/models/credential/__init__.py:
```diff
@@ -1257,3 +1307,72 @@ def create(self):
+
+    target_credential = models.ForeignKey(
+        'Credential',
+        related_name='input_source',
```
`'input_source'` -> `'input_sources'`

It's almost impossible to fix these after we get the code merged, so I'd like to get them fixed in review.

consider `input_source for input_source in credential.input_source.all()`

vs.

`input_source for input_source in credential.input_sources.all()`

ryanpetrello replied:
Good 👁

AlanCoding commented on awx/main/models/credential/__init__.py:
```diff
@@ -1257,3 +1307,72 @@ def create(self):
+    )
+    source_credential = models.ForeignKey(
+        'Credential',
+        related_name='target_input_source',
```
same for target_input source

AlanCoding commented on awx/main/models/credential/__init__.py:
```diff
@@ -778,6 +823,11 @@ def create(self):
+            'id': 'ssh_public_key_data',
+            'label': ugettext_noop('Signed SSH Certificate'),
+            'type': 'string',
+            'multiline': True,
```
Is this also addressing https://github.com/ansible/awx/issues/1654?

The implementation of certificates here seems different from https://github.com/ansible/awx/issues/2434, it would be good to get a disposition on how that would be done after this is merged.

ryanpetrello replied:
Yes, this supports #1654 (I wasn't aware that issue existed).

This implementation supports *both* use cases.

You could _either_ paste in your **already signed** data _directly_ and `ssh-add` will just do the right thing, *or* you can paste in **unsigned** data and associate a HashiCorp SSH Signer input source for signing purposes.  Both work.

ryanpetrello replied:
@AlanCoding from speaking with @wenottingham, 
https://github.com/ansible/awx/issues/2434 is tracking a _generalized_ enhancement to provide a built-in credential type where we'd support certificate-based auth _to cloud services_, **not** SSH).  @jakemcdermott and I do not intend to solve _that_ use case in this PR.

**This** PR (in part) solves two use cases of "I want to do SSH certificate auth"

## User Brings a Private Key and Signed Cert

Under this model, an AWX user would:

1.  Generate a key and signed cert (or use an existing one).  They might use some automated tooling to generate a signed certificate, or they might do it by hand.  Either way, they'll have e.g,. an `id_rsa` and an `id_rsa-cert.pub`.
2.  Put the private key and certificate into keys/buckets/whatever in a `<Azure/HashiCorp/whatever>` key-value store.
3.  Configure a Machine credential in AWX that:
   a. pulls the `ssh_key_data` (`id_rsa`) *out* of the secret store
   b. pulls the `ssh_public_key_data` (`id_rsa-cert.pub`) *out* of the secret store
4.  Code in this PR would run `ssh-add` appropriately such that these values are utilized for playbook runs.

## User Brings a Private Key and HashiVault Signs the Public Half

Under this model, an AWX user would:

1.  Call `ssh-keygen` to generate a key (or use an existing one).  They'll have e.g,. an `id_rsa` and an unsigned `id_rsa.pub`.
2.  Put the private key (`id_rsa`) into HashiCorp Vault's key-value store.
3.  Configure a Machine credential in AWX that:
   a. pulls the `ssh_key_data` *out* of HashiCorp Vault
   b. stores the _unsigned public half_ (`id_rsa.pub`) in the AWX database under `ssh_public_key_data`, and configures HashiCorp Vault to _sign_ that value (_when the playbook runs_) using some role they've defined in HashiCorp Vault
4.  Code in this PR would run `ssh-add` appropriately such that the private key and the _signed_ cert from HashiCorp Vault were utilized for playbook runs and lots of users.

This second model also supports the configurables you see under https://www.vaultproject.io/api/secret/ssh/index.html#sign-ssh-key for SSH signing (things like specifying principals, and even TTLs).  Under this model, for example, you could have Vault hand out certificates with short (30s) expirations.  This model also vastly simplifies revocation for organizations with really large infrastructure.

matburt replied:
👍

jakemcdermott replied:
lgtm 👍 we tested signing and revocation via TTL yesterday for the second model

jakemcdermott committed:
add input source prompting and plugin testing

ryanpetrello committed:
add API examples for supported credential plugins

ryanpetrello committed:
Encrypt machine.ssh_public_key_data (in case users paste in signed data)

ryanpetrello committed:
rename the CredentialInputSource related_names so they're plural

ryanpetrello committed:
allow creation at /api/v2/credential_input_sources

jakemcdermott committed:
cache dynamic input fields

Query dynamic input fields once on attribute access and then cache it for future use.

ryanpetrello committed:
remove association behavior from /api/v2/credentials/input_sources/

AlanCoding commented on awx/api/serializers.py:
```diff
@@ -2801,6 +2802,44 @@ def create(self, validated_data):
+        }
+
+    def get_source_credential_name(self, obj):
+        return obj.source_credential.name
```
these 2 linked credentials don't show up in summary fields?

I'm worried that people might make subsequent requests to show them in summary fields.

ryanpetrello replied:
@jakemcdermott I think I agree with this; it's kind of abnormal for us to add distinct _fields_ like this for things like e.g., `credential_name`, `credential_type` etc....  This is the type of data that we *generally* put in `summary_fields`.  We should probably have a `summary_fields` on this endpoint that shows ID, name, maybe some other fields for the source and target credential.

What do you think?

https://github.com/ansible/awx/pull/3334/

jakemcdermott replied:
works for me

ryanpetrello committed:
add related and summary fields to the CredentialInputSource endpoint

AlanCoding commented on awx/main/models/credential/__init__.py:
```diff
@@ -363,6 +372,14 @@ def passwords_needed(self):
+        if dynamic_input_fields is None:
+            self._dynamic_input_fields = self._get_dynamic_input_field_names()
+            return self._dynamic_input_fields
+        return dynamic_input_fields
```
The way that you're handing the internal property looks like this is a vanilla cached property. You might want to have a look at `from django.utils.functional import cached_property`, it's just a way to avoid having to maintain that overhead yourself. But this done in a million places, and its current form is fine with me.

Thanks for getting rid of the overrides of `__init__` and other methods like that, this is looking much better.

jakemcdermott replied:
@AlanCoding Neat! I didn't know that existed. This reminds me of the classic cache decorator [covered in a really great pycon talk a few years back](https://speakerdeck.com/pyconslides/transforming-code-into-beautiful-idiomatic-python-by-raymond-hettinger-1?slide=33). I thought about adding something similar for this, but I didn't want to add a new helper function for a single use case.

As you've pointed out, django provides something comparable in its library of utility functions so we can just use that 👍

AlanCoding commented on docs/credential_plugins.md:
```diff
@@ -0,0 +1,329 @@
+the Machine Credential.
+
+1.  Create the Machine Credential with a static username, `example-user`.
+
```
Could you clarify what to expect in the case that a field is required?

In the case of this example, say that instead of the "Machine" type, I'm using "Amazon Web Services". I want to use hashicorp vault to store the `password` field.

For AWS credentials, `password` is a required field. So in Step (1) here, I cannot create the credential without providing a value for the password. Do I have to put in `"password": "foo"`, so that the API will let me save it so that I can continue through the next steps to link to the external secrets?

ryanpetrello replied:
Bleh, this is a good question, and I don't have a good answer off the top of my head.  We may need to update some of these to remove the `required` requirement (since you can't really apply the input source relationship until the object itself has been saved).

ryanpetrello replied:
Okay, addressed in https://github.com/ansible/awx/pull/3098/commits/7d4c32f38c0f1cfa155c80f18d284818ee68e797

AlanCoding commented on awx/main/models/credential/__init__.py:
```diff
@@ -1257,3 +1312,80 @@ def create(self):
+
+    def save(self, *args, **kwargs):
+        self.full_clean()
+        super(CredentialInputSource, self).save(*args, **kwargs)
```
Can we just delete this method? Our custom `BaseSerializer` is explicitly designed to create a mock object and call all the `clean_` methods as a part of the validation. I don't see any reason that `full_clean` is needed to be called here, and it's a suspicious departure from standard patterns.

AlanCoding commented on awx/main/models/credential/__init__.py:
```diff
@@ -461,8 +480,18 @@ def get_input(self, field_name, **kwargs):
+
+    def _get_dynamic_input(self, field_name):
+        input_sources = CredentialInputSource.objects.filter(target_credential=self)
+        return input_sources.filter(input_field_name=field_name).first().get_input_value()
```
This is an extra query, and it surfaces errors in a confusing way ("NoneType has no attribute `get_input_value`").

```python
for input_source in self.input_sources.all():
    if input_source.input_field_name == field_name:
        return input_source.get_input_value()
else:
    raise RuntimeError('Field name {} is not a dynamic input field, choices are {}...")
```

I just want to throw that out there, and I know it feels like more code, but it avoids extra queries if you can go through the effort to prefetch the `input_sources`.

AlanCoding commented on awx/main/access.py:
```diff
@@ -1125,6 +1125,46 @@ def can_delete(self, obj):
+     - I'm an admin of the associated target_credential
+    '''
+
+    model = CredentialInputSource
```
you could add a `select_related` or `prefetch_related` for the `target_credential` and `source_credential`. That will make the listings of these go from 🐢 to 🐇

AlanCoding commented on awx/main/models/credential/__init__.py:
```diff
@@ -441,6 +458,8 @@ def get_input(self, field_name, **kwargs):
         :param field_name(str):        The name of the input field.
         :param default(optional[str]): A default return value to use.
         """
+        if field_name in self.dynamic_input_fields:
```
For the sake of my dark paranoia, could you add to this condition:

```python
if self.kind != 'external' and field_name in self.dynamic_input_fields:
```

And then document that external credential sources _cannot_ ever be used to look up secrets for external credential sources. And everyone who had this thought should feel bad for thinking it.

ryanpetrello committed:
clean up access to various CredentialInputSource fields (#3336)

jakemcdermott committed:
document restriction of external-external credential source linking

AlanCoding commented:
In cases of referencing the list of a job's credentials in the task system:

https://github.com/ansible/awx/blob/cebd918e49fd93b47d49ec934a1702ee06a8162c/awx/main/tasks.py#L1157

https://github.com/ansible/awx/blob/cebd918e49fd93b47d49ec934a1702ee06a8162c/awx/main/tasks.py#L1012

I would suggest using `job.credentials.prefetch_related('input_sources__source_credential').all()` as an optimization, because it is expected that `get_input` will be called multiple times for the same stuff by our task logic, without consideration of performance.

I figure that you'll probably be running some jobs before finalizing your branch for merge, and this will be a nice thing to piggyback on that testing.

ryanpetrello committed:
enforce required credential fields at job start time rather than on save

this is necessary for credential plugins support so that you can (in two
requests):

1.  Save a Credential with _no_ input values defined
2.  Create/associate one (or more) CredentialInputSource records to the
    new Credential

ryanpetrello committed:
prefetch related source credentials in tasks.py

ryanpetrello committed:
add support for CyberArk Conjur (API v5)

jakemcdermott committed:
support input source metadata in plugin test apis

ryanpetrello committed:
store the public key for HashiVault signing in the plugin metadata

AlanCoding commented on awx/main/migrations/0064_v350_credential_plugins.py:
```diff
@@ -0,0 +1,53 @@
+
+    dependencies = [
+        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
+        ('taggit', '0002_auto_20150616_2121'),
```
Why would we add a 0002 migration dependency, as opposed to deleting the taggit app? And why do either of these things now?

ryanpetrello replied:
That's a great question.  If you can explain to me why Django automatically adds these two dependencies when I run `awx-manage makemigrations`, points for you :)

AlanCoding replied:
It looks like we already have the taggit 0002 migration as a dependency in our `0001_initial` migration.

I would think that our very own 0001 migration would be an ancestor of this migration, ergo, taggit 0002 is already an implicit dependency. It didn't do this for me when I added 1 simple field.

Also, just to re-state the general situation, I know of literally no reason to have the taggit app.

AlanCoding replied:
well, it seems pretty clear that it did this for your migration because you have references to the auth app's User model and the taggit app's tag bonanza. For minor field changes, this is not relevant.

Maybe Django wants to add in redundant dependency links for all these cases, because they wouldn't do any harm.

AlanCoding commented on awx/main/tests/functional/conftest.py:
```diff
@@ -293,6 +329,18 @@ def org_credential(organization, credentialtype_aws):
+@pytest.fixture
+def other_external_credential(credentialtype_external):
+    return Credential.objects.create(credential_type=credentialtype_external, name='other-external-cred',
+                                     inputs={'url': 'http://testhost.com', 'token': 'secret2'})
```
The fixture flow for these is very clear, and these are things we want to have ready to go in the fixture namespace 👍 

The only other thing I would prefer to add is a `credential_input_source` fixture (or whatever the most canonical name would be). A lot of your tests seem to be really long and linear, because you tend to create the input sources through the mock API first. Those would be more clear if they could just start out with the object taken from a generic fixture.

AlanCoding reviewed (approved):
I've focused on the API views and models and task implications, and I'm finished with reviewing. All the important points I made have been addressed.

jakemcdermott committed:
add cyberark conjur to tested credential types

jakemcdermott committed:
add input source creation ui

jakemcdermott committed:
move input value initialization to models

jakemcdermott committed:
move org edit permission check to route resolve

jakemcdermott committed:
combine add and edit controllers

jakemcdermott committed:
add validation for required values and metadata fields

jakemcdermott committed:
don't show lookup until data is fetched

jakemcdermott commented:
recheck

jakemcdermott commented:
recheck

jakemcdermott committed:
add selected credential tray to input source lookup

jakemcdermott committed:
add wrapper for text input tags

jakemcdermott committed:
enable input source linking for password fields

jakemcdermott commented:
![replace_revert](https://user-images.githubusercontent.com/9753817/53897780-d6d76d00-4004-11e9-83e5-d38e8a18e0bb.gif)

related: - https://github.com/ansible/awx/issues/2834

jakemcdermott committed:
disable prompt-on-launch when input source is set

jakemcdermott committed:
enable input source linking for textarea fields

jakemcdermott committed:
enable input source linking for secret textarea fields

kialam commented on awx/ui/client/lib/components/input/textarea-secret.directive.js:
```diff
@@ -42,9 +42,13 @@ function AtInputTextareaSecretController (baseInputController, eventService) {
         scope.$watch('state[state._activeModel]', () => vm.check());
         scope.$watch('state._isBeingReplaced', () => vm.onIsBeingReplacedChanged());
+
+        console.log(scope.state._value);
```
👀

jakemcdermott replied:
: D

jakemcdermott committed:
don't replace input source unless changed

jakemcdermott committed:
add tips for secret controls

jlmitch5 commented on awx/ui/client/lib/components/tag/_index.less:
```diff
@@ -8,6 +8,7 @@
     flex-direction: row;
     align-content: center;
     min-height: @at-space-4x;
+    max-height: @at-space-4x;
```
what if tag names are long and take up multiple lines?

jakemcdermott replied:
They'll get truncated - I think they should show ellipses when this happens but they don't currently, what do you think?

jakemcdermott replied:
At a minimum, we don't want this to apply to all tags so I've moved this change to the wrapper class used for input field tags.

jlmitch5 commented on awx/ui/client/lib/components/lookup-list/lookup-list.partial.html:
```diff
@@ -0,0 +1,90 @@
+            <div class="List-lookupLayout List-tableHeaderRow">
+                <div></div>
+                <div class="d-flex h-100">
+                    <div
```
this looks as if it is being passed a number of isolate scope attributes but it is a div (not an angular directive/component?)

jakemcdermott replied:
I agree with you. I generated the template using our state generator and list generator code as a starting point and then used it to create a somewhat portable component that could be embedded inside of a modal and explicitly controlled from one place.

The attributes seem to be needed for sorting to work. I think they are related our legacy list / table code and styles, search, pagination, etc., which makes me hesitant to change it right now. If you were to generate a lookup in the usual way and then inspect the resultant template, you would see these there, too.

jakemcdermott replied:
We talked about this one - given what I've mentioned above, we're going to leave this as is.

jlmitch5 commented on awx/ui/client/lib/components/input/textarea.partial.html:
```diff
@@ -1,17 +1,40 @@
-                  ng-change="vm.check()"
-                  ng-disabled="state._disabled || form.disabled" /></textarea>
-
+        <div ng-class="{ 'input-group': state.tagMode }">
```
this to me reads as if it was a component for textarea...the component seems to have shifted to be more like our lookup (with a button to the left...just with a text area instead of an input).  It is confusing for this be named textarea, based on what I understand.  Are there other forms that are using the textarea component?  (doubt it, but just a note it would be good to check).

jakemcdermott replied:
I see what you mean. I added this whole div to the textarea components for a very specific state where we need make the field _look_ like a tag inside of a text area:

<img width="1118" alt="Screen Shot 2019-03-14 at 12 19 30 PM" src="https://user-images.githubusercontent.com/9753817/54373337-74f5b380-4653-11e9-89af-07e17cc04428.png">

I didn't find a way of doing this by nesting a tag component inside of the actual `textarea` element, so when we need to show the component in "tag mode", we're actually just hiding the entire textarea and showing this div instead.

There are no other forms using this component. I'm open to any specific suggestions you might have here - are you proposing that we rename the component to make it clear that it is customized specifically for the credentials form?

jakemcdermott replied:
We talked about this offline. The improvement @jlmitch5  suggested that I'm going to make here is to update the hint text when the component is in tag mode so that it doesn't suggest that you can drag and drop or enter text (because you can't). Instead, it will give a different hint to the user that they must first unlink the input source (remove the tag) if they wish to drag-and-drop an actual ssh key.

jakemcdermott replied:
@jlmitch5 I've updated the textarea components so that they don't show the drag-and-drop hint when the input value is being displayed as a tag.

jlmitch5 commented on awx/ui/client/lib/components/input/text.partial.html:
```diff
@@ -1,15 +1,42 @@
+            ng-disabled="state._disabled || form.disabled"
+          />
+        </div>
+        <input ng-if="!state.tagMode" type="text" class="form-control at-Input"
```
the way this input is set up makes a bit more sense to me...standard input is given if tagMode is not passed.

jakemcdermott replied:
👍 That is good feedback to have, thank you. For what it is worth, the actual behavior of the textarea components discussed above is meant to be the same as it is here: The textarea should behave as a a standard textarea component unless `_tagMode` is passed.

jlmitch5 commented on awx/ui/client/lib/components/easy-modal/easy-modal.component.js:
```diff
@@ -0,0 +1,29 @@
+const templateUrl = require('~components/easy-modal/easy-modal.partial.html');
```
I would probably just name this dialog, as that what it seems to be (a modal with just a simple text body content)

jakemcdermott replied:
It is being used for more than just text content. Its meant to encase anything you want to put in it, including other components.

We use it to display the plugin lookup and input prompt here: https://github.com/ansible/awx/blob/credential_plugins/awx/ui/client/features/credentials/input-source-lookup.partial.html#L1

We also use it for the external test prompt we display when the test button is clicked on an external credential: https://github.com/ansible/awx/blob/credential_plugins/awx/ui/client/features/credentials/external-test-modal.partial.html#L1

With that in mind, would you still like me to rename it?

jakemcdermott replied:
Sounds good to me - I've renamed the component to `dialog.component.js`

jlmitch5 commented on awx/ui/client/lib/components/action/action-button.partial.html:
```diff
@@ -0,0 +1,3 @@
+<button class="btn at-Button{{ vm.fill }}--{{ vm.color }}">
```
change to {{ vm.color ? `--${vm.color}` : '' }} and this will pass the working at-Button class when neither fill or color are passed

jakemcdermott replied:
👍I will do that.

kialam commented on awx/ui/client/features/credentials/_index.less:
```diff
@@ -0,0 +1,21 @@
+.InputSourceLookup-selectedItemLabel {
+    color: @default-interface-txt;
+    text-transform: uppercase;
+    margin-right: 10px;
```
We have pre-defined variables to handle layout/whitespace declarations in `_variables.less`. You can write the following two lines as `margin: 0 @at-space-2x;`

jakemcdermott replied:
good idea - done.

kialam commented on awx/ui/client/features/credentials/add-edit-credentials.controller.js:
```diff
@@ -0,0 +1,610 @@
+     * metadata object that can be sent to the api later or reloaded when re-opening the form.
+     */
+    function getMetadataFormSubmitData ({ inputs }) {
+        const metadata = Object.assign({}, ...inputs._group
```
Instead of chaining `.filter`, `.map`, and `Object.assign` to craft the final form of the array you are processing, I think a `.reduce` method here fits the bill nicely because it will do all three in one go, and is meant for these types of conversions:

```
function getMetadataFormSubmitData ({ inputs }) {
        return inputs._group
            .reduce((formatted, { id, _value }) => {
                if (_value) {
                    formatted[id] = _value;
                }
                return formatted;
            }, {});
        } 
```

jakemcdermott replied:
👍 works for me

jakemcdermott committed:
use a shared variable for layout declarations

jakemcdermott committed:
move tag max height declaration to input tag wrapper

We don't want to apply max height to all tags, just the ones we embed
within text/textarea input fields.

jakemcdermott committed:
refactor metadata conversion function to use reduce

jakemcdermott committed:
use default action button class when fill and color props aren't given

jakemcdermott committed:
rename at-easy-modal to at-dialog

jakemcdermott committed:
remove bootstrap fade from dialog component

This was breaking the fixed positioning of help popovers for form inputs
nested within the dialog so I'm replacing it.

jakemcdermott committed:
hide drag-and-drop textarea hints when showing input value as tag

You can't drag and drop or enter text when a text area is displaying its value
as a tag, so we hide the hint to avoid incorrectly suggesting that you can.

jakemcdermott committed:
remove non-matches from input source arrays

jakemcdermott committed:
add plugin for cyberark aim

jakemcdermott committed:
read cert from memory

Certificate data lives in the db. We're using a fifo here so we can avoid writing
it to disk when making the http request.

jakemcdermott committed:
generalize secret textarea hint

jakemcdermott committed:
don't ignore choices param in input config

We always want to use choices if they're available in the input config
from the api. An input can sometimes have no type. Usually we'd throw an
error but we can still load a component if it defines a set of choices
to use instead of a type.

jakemcdermott committed:
add object query fields

ryanpetrello commented on docs/credential_plugins.md:
````diff
@@ -0,0 +1,315 @@
+```python
+from awx.main.credential_plugins import hashivault
+hashivault.hashivault_kv_plugin.backend(
+    '',
````
@jakemcdermott can you remove this line?

jakemcdermott committed:
remove input value arg from plugin example

The plugin backends no longer expect an input value as an argument
so we're removing it from the documented examples.

jakemcdermott committed:
add use to credential user_capabilities

jakemcdermott committed:
add delete to input source user_capabilities

jakemcdermott committed:
disable input source lookups when they can't be changed

If a user doesn't have permission to change an input field, we disable
the input field.

jakemcdermott commented on awx/main/access.py:
```diff
@@ -1162,6 +1162,52 @@ def can_delete(self, obj):
 
+    def get_user_capabilities(self, obj, **kwargs):
+        user_capabilities = super(CredentialAccess, self).get_user_capabilities(obj, **kwargs)
+        user_capabilities['use'] = self.can_use(obj)
```
Assuming you have sufficient permissions on the target credential, you need `use` or better permissions on a source credential in order to link it. 

By including `use` in the `user_capabilities` object, we can know if an external credential is linkable by inspecting its summary_fields.

jakemcdermott committed:
update validation when input source is set

jakemcdermott committed:
don't initialize linked textarea fields as invalid

jakemcdermott committed:
add workaround for null cloud credential kind summary fields

job details summary field credentials have `null` for the
`kind` field so we can't map it to icon names. To work around
this, we now check the value of the `cloud` field when generating
the icon to use.

jakemcdermott committed:
remove linked fields from validation loop

All linked fields are initially valid. All unlinked fields
are initially valid.

jakemcdermott committed:
add timeouts to plugin backends

jakemcdermott committed:
prevent overflow in plugin test notifications

jakemcdermott committed:
prevent input source changes without use role on source cred

To update an input source, the user must have admin access
to the target credential and at least use role on the source
credential.

jakemcdermott committed:
include externally sourced input fields for injection

jakemcdermott commented on awx/main/models/credential/__init__.py:
```diff
@@ -653,15 +698,15 @@ class TowerNamespace:
         # DB storage)
-        for field_name, value in credential.inputs.items():
+        injectable_fields = list(credential.inputs.keys()) + credential.dynamic_input_fields
+        for field_name in list(set(injectable_fields)):
```
If a field has a value assigned to it normally, it can be found in `credential.inputs`. If it's sourced from an external credential, it'll be in `credential.dynamic_input_fields`. We need to include both when injecting values.

jakemcdermott committed:
fix required field entry for object_query

jakemcdermott committed:
make linked fields look disabled when disabled

jakemcdermott committed:
render non-editable linked credentials

We still want to show linked credentials even when a field is
non-editable.

jakemcdermott committed:
update e2e credential checks

jakemcdermott committed:
replace unlinked fields before updating input sources

With credential type changes for linked credentials disabled, we can
update the unlinked credential fields first. If an error occurs when
updating the unlinked fields, no input sources will be changed.

jakemcdermott committed:
mention usage with custom credential types in credential plugins doc

sumkincpp commented:
Wow, this feature really worth having, really nice that it's merged!

robison commented:
This is really phenomenal. Great work.

kevin-dfo commented:
This is cool.  Would it be possible to use a variable in the metadata? ie. Path to secret: /secret/{{ inventory_hostname }} to get per host username/passwords?

Or is the method mentioned here https://www.ansible.com/blog/ansible-tower-feature-spotlight-custom-credentials still the best way for that?

jakemcdermott commented:
@kevin-dfo We may introduce a way to define metadata dynamically at a future point in time, but for right now the method mentioned in the blog post you linked is still the best way for doing that.

bbaassssiiee commented:
It seems that the `hvac` python library is missing, this is required for HashiCorp Vault lookup.

ryanpetrello commented:
@bbaassssiiee this integration does not make use of https://docs.ansible.com/ansible/latest/plugins/lookup/hashi_vault.html
